### PR TITLE
gh-117531: Unblock getters after non-immediate queue shutdown

### PR DESCRIPTION
**Pull request by EpicWink**
* Issue: gh-117531

### CODE REVIEW

File: Lib/test/test_queue.py [hunk 1]
```diff
@@ -636,6 +636,23 @@ def test_shutdown_get_task_done_join(self):
 
         self.assertEqual(results, [True]*len(thrds))
 
+    def test_shutdown_get_simple(self):
```
Comment (YvesDup) [inline Lib/test/test_queue.py]: `def test_shutdown_pending_get` seems to me more explicit.

File: Lib/test/test_queue.py [hunk 2]
```diff
@@ -636,6 +636,23 @@ def test_shutdown_get_task_done_join(self):
...
+            try:
+                results.append(q.get())
+            except Exception as e:
+                results.append(e)
+
+        q = self.type2test()
+        results = []
+        get_thread = threading.Thread(target=get)
+        get_thread.start()
+        q.shutdown()
```
Comment (YvesDup) [inline Lib/test/test_queue.py]: I suggest: `q.shutdown(immediate=False)`.

Should we add a test when shutdown is immediate (True) ? 
And should we add a test to check pending put ?

Comment (EpicWink) [reply Lib/test/test_queue.py]: > Should we add a test when shutdown is immediate (True) ?
> And should we add a test to check pending put ?

The bug is specifically for getters when `immediate=False`, no need to increase the scope of the PR.

---

> I suggest: q.shutdown(immediate=False).

Good idea

Comment (YvesDup) [reply Lib/test/test_queue.py]: > The bug is specifically for getters when immediate=False, no need to increase the scope of the PR.

This seems reasonable. 
I approve the changes.

**Comment by YvesDup** [comment]
Regarding [my last comments](https://github.com/python/cpython/pull/104228/files#r1554646081), docstring of `shutdown` method (and related documentation - I didn't check) must be modified about  'blocked callers'.

**Comment by EpicWink** [comment]
> Regarding [my last comments](https://github.com/python/cpython/pull/104228/files#r1554646081), docstring of shutdown method (and related documentation - I didn't check) must be modified about 'blocked callers'.

Not relevant for this PR, perhaps you meant to comment on #104750

File: Lib/queue.py [hunk 3]
```diff
@@ -253,6 +253,7 @@ def shutdown(self, immediate=False):
                 # release all blocked threads in `join()`
                 self.all_tasks_done.notify_all()
             self.not_empty.notify_all()
+            # All getters need to re-check queue-empty to raise ShutDown
```
Comment (YvesDup) [inline Lib/queue.py:255]: It seems to me better if you move up this new comment above the line 255 ?

**Comment by YvesDup** [comment]
I approved the last changes

**Comment by YvesDup** [comment]
I approve

**Comment by YvesDup** [comment]
@gvanrossum Sorry, I forget to tell you that I've approved

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**